nativelib/src/main/resources/gc/commix/Heap.c

```diff
@@ -0,0 +1,308 @@
+    GreyList_PushAll(&heap->mark.empty, greyPacketsStart,
+                     (GreyPacket *)greyPacketsStart, greyPacketCount);
+
+    // reserve space for bytemap
+    Bytemap *bytemap = (Bytemap *)Heap_mapAndAlign(
+        maxHeapSize / ALLOCATION_ALIGNMENT + sizeof(Bytemap),
+        ALLOCATION_ALIGNMENT);
+    heap->bytemap = bytemap;
+
+    // Init heap for small objects
+    heap->heapSize = minHeapSize;
+    heap->heapStart = heapStart;
+    heap->heapEnd = heapStart + minHeapSize / WORD_SIZE;
+
+    Phase_Init(heap, initialBlockCount);
+
+    Bytemap_Init(bytemap, heapStart, maxHeapSize);
+    Allocator_Init(&allocator, &blockAllocator, bytemap, blockMetaStart,
+                   heapStart);
+
+    LargeAllocator_Init(&largeAllocator, &blockAllocator, bytemap,
+                        blockMetaStart, heapStart);
+
+    // Init all GCThreads
+    // Init stats if enabled.
+    // This must done before initializing other threads.
+    heap->stats = Stats_OrNull(Heap_createMutatorStats());
+
+    int gcThreadCount = Settings_GCThreadCount();
+    heap->gcThreads.count = gcThreadCount;
+    Phase_Set(heap, gc_idle);
+    GCThread *gcThreads = (GCThread *)malloc(sizeof(GCThread) * gcThreadCount);
+    heap->gcThreads.all = (void *)gcThreads;
+    for (int i = 0; i < gcThreadCount; i++) {
+        Stats *stats = Stats_OrNull(Heap_createStatsForThread(i));
+        GCThread_Init(&gcThreads[i], i, heap, stats);
+    }
+
+    heap->mark.lastEnd_ns = scalanative_nano_time();
+
+    pthread_mutex_init(&heap->sweep.growMutex, NULL);
+}
+
+void Heap_Collect(Heap *heap) {
+    Stats *stats = Stats_OrNull(heap->stats);
+    Stats_CollectionStarted(stats);
+    assert(Sweeper_IsSweepDone(heap));
+#ifdef DEBUG_ASSERT
+    Sweeper_ClearIsSwept(heap);
+    Sweeper_AssertIsConsistent(heap);
+#endif
+    Phase_StartMark(heap);
+    Marker_MarkRoots(heap, stats);
+    Marker_MarkUtilDone(heap, stats);
+    Phase_MarkDone(heap);
+    Stats_RecordEvent(stats, event_mark, heap->mark.currentStart_ns,
+                      heap->mark.currentEnd_ns);
+    Phase_StartSweep(heap);
+}
+
+bool Heap_shouldGrow(Heap *heap) {
+    uint32_t freeBlockCount = (uint32_t)blockAllocator.freeBlockCount;
+    uint32_t blockCount = heap->blockCount;
+    uint32_t recycledBlockCount = (uint32_t)allocator.recycledBlockCount;
+    uint32_t unavailableBlockCount =
+        blockCount - (freeBlockCount + recycledBlockCount);
+
+#ifdef DEBUG_PRINT
+    printf("\n\nBlock count: %" PRIu32 "\n", blockCount);
+    printf("Unavailable: %" PRIu32 "\n", unavailableBlockCount);
+    printf("Free: %" PRIu32 "\n", freeBlockCount);
+    printf("Recycled: %" PRIu32 "\n", recycledBlockCount);
+    fflush(stdout);
+#endif
+
+    uint64_t timeInMark = heap->mark.currentEnd_ns - heap->mark.currentStart_ns;
+    uint64_t timeTotal = heap->mark.currentEnd_ns - heap->mark.lastEnd_ns;
+
+    return timeInMark >= GROWTH_MARK_FRACTION * timeTotal ||
+           freeBlockCount * 2 < blockCount ||
```

Comment:
Free block fraction is 50%, lets make it a publicly configurable setting, similarly to heap size.

Reply:
i.e., `SCALANATIVE_FREE_RATIO`.

Reply:
Done